File: packages/emitter/testcases/js-only/real-world/data-structures/data-structures.ts
export class Stack<T> {
  private items: T[] = [];

  push(item: T): void {
    this.items.push(item);
  }

  // Idiomatic C# pattern: TryPop with out parameter
  tryPop(result: out<T>): boolean {
    if (this.items.length === 0) {
      return false;
    }
    result = this.items.pop()!;
    return true;
  }

  // Idiomatic C# pattern: TryPeek with out parameter
  tryPeek(result: out<T>): boolean {
    if (this.items.length === 0) {
      return false;
    }
    result = this.items[this.items.length - 1];
    return true;
  }

  isEmpty(): boolean {
    return this.items.length === 0;
  }

  size(): number {
    return this.items.length;
  }

  clear(): void {
    this.items = [];
  }
}

export class Queue<T> {
  private items: T[] = [];

  enqueue(item: T): void {
    this.items.push(item);
  }

  // Idiomatic C# pattern: TryDequeue with out parameter
  tryDequeue(result: out<T>): boolean {
    if (this.items.length === 0) {
      return false;
    }
    result = this.items.shift()!;
    return true;
  }

  // Idiomatic C# pattern: TryPeek with out parameter
  tryPeek(result: out<T>): boolean {
    if (this.items.length === 0) {
      return false;
    }
    result = this.items[0];
    return true;
  }

  isEmpty(): boolean {
    return this.items.length === 0;
  }

  size(): number {
    return this.items.length;
  }

  clear(): void {
    this.items = [];
  }
}

export class LinkedListNode<T> {
  constructor(
    public value: T,
    public next: LinkedListNode<T> | null = null
  ) {}
}

export class LinkedList<T> {
  private head: LinkedListNode<T> | null = null;
  private tail: LinkedListNode<T> | null = null;
  private length: number = 0;

  append(value: T): void {
    const node = new LinkedListNode(value);

    if (this.head === null) {
      this.head = node;
      this.tail = node;
    } else {
      if (this.tail) {
        this.tail.next = node;
      }
      this.tail = node;
    }

    this.length++;
  }

  prepend(value: T): void {
    const node = new LinkedListNode(value, this.head);
    this.head = node;

    if (this.tail === null) {
      this.tail = node;
    }

    this.length++;
  }

  size(): number {
    return this.length;
  }

  toArray(): T[] {
    const result: T[] = [];
    let current = this.head;

    while (current !== null) {
      result.push(current.value);
      current = current.next;
    }

    return result;
  }
}

export function testDataStructures(): void {
  const stack = new Stack<number>();
  stack.push(1);
  stack.push(2);
  stack.push(3);
  console.log("Stack size:", stack.size());

  let poppedValue: number;
  if (stack.tryPop(poppedValue)) {
    console.log("Stack pop:", poppedValue);
  }

  const queue = new Queue<string>();
  queue.enqueue("first");
  queue.enqueue("second");
  queue.enqueue("third");
  console.log("Queue size:", queue.size());

  let dequeuedValue: string;
  if (queue.tryDequeue(dequeuedValue)) {
    console.log("Queue dequeue:", dequeuedValue);
  }

  const list = new LinkedList<number>();
  list.append(10);
  list.append(20);
  list.prepend(5);
  console.log("List as array:", list.toArray());
}

File: packages/emitter/testcases/js-only/real-world/data-structures/data-structures.cs
namespace TestCases.jsonly.realworld.datastructures
{
    public class Stack<T>
    {
        private global::System.Collections.Generic.List<T> items = new global::System.Collections.Generic.List<T>();

        public void push(T item)
            {
            global::Tsonic.JSRuntime.Array.push(this.items, item);
            }

        public bool tryPop(out T result)
            {
            result = default;

            if (global::Tsonic.JSRuntime.Array.length(this.items) == 0)
                {
                return false;
                }
            result = this.items.pop()!;
            return true;
            }

        public bool tryPeek(out T result)
            {
            result = default;

            if (global::Tsonic.JSRuntime.Array.length(this.items) == 0)
                {
                return false;
                }
            result = global::Tsonic.JSRuntime.Array.get(this.items, global::Tsonic.JSRuntime.Array.length(this.items) - 1);
            return true;
            }

        public bool isEmpty()
            {
            return global::Tsonic.JSRuntime.Array.length(this.items) == 0;
            }

        public double size()
            {
            return global::Tsonic.JSRuntime.Array.length(this.items);
            }

        public void clear()
            {
            this.items = new global::System.Collections.Generic.List<T>();
            }
    }
    public class Queue<T>
    {
        private global::System.Collections.Generic.List<T> items = new global::System.Collections.Generic.List<T>();

        public void enqueue(T item)
            {
            global::Tsonic.JSRuntime.Array.push(this.items, item);
            }

        public bool tryDequeue(out T result)
            {
            result = default;

            if (global::Tsonic.JSRuntime.Array.length(this.items) == 0)
                {
                return false;
                }
            result = this.items.shift()!;
            return true;
            }

        public bool tryPeek(out T result)
            {
            result = default;

            if (global::Tsonic.JSRuntime.Array.length(this.items) == 0)
                {
                return false;
                }
            result = global::Tsonic.JSRuntime.Array.get(this.items, 0);
            return true;
            }

        public bool isEmpty()
            {
            return global::Tsonic.JSRuntime.Array.length(this.items) == 0;
            }

        public double size()
            {
            return global::Tsonic.JSRuntime.Array.length(this.items);
            }

        public void clear()
            {
            this.items = new global::System.Collections.Generic.List<T>();
            }
    }
    public class LinkedListNode<T>
    {
        public T value;

        public LinkedListNode<T>? next;

        public LinkedListNode(T value, LinkedListNode<T>? next = null)
            {
            this.value = value;
            this.next = next;
            }
    }
    public class LinkedList<T>
    {
        private LinkedListNode<T>? head = null;

        private LinkedListNode<T>? tail = null;

        private double length = 0;

        public void append(T value)
            {
            var node = new LinkedListNode(value);
            if (this.head is null)
                {
                this.head = node;
                this.tail = node;
                }
            else
                {
                if (this.tail != null)
                    {
                    this.tail.Next = node;
                    }
                this.tail = node;
                }
            this.length++;
            }

        public void prepend(T value)
            {
            var node = new LinkedListNode(value, this.head);
            this.head = node;
            if (this.tail is null)
                {
                this.tail = node;
                }
            this.length++;
            }

        public double size()
            {
            return this.length;
            }

        public global::System.Collections.Generic.List<T> toArray()
            {
            global::System.Collections.Generic.List<T> result = new global::System.Collections.Generic.List<T>();
            var current = this.head;
            while (current is not null)
                {
                global::Tsonic.JSRuntime.Array.push(result, current.Value);
                current = current.Next;
                }
            return result;
            }
    }

            public static class datastructures
            {
                public static void testDataStructures()
                    {
                    var stack = new Stack<double>();
                    stack.Push(1);
                    stack.Push(2);
                    stack.Push(3);
                    global::Tsonic.JSRuntime.console.log("Stack size:", stack.size());
                    double poppedValue = 0;
                    if (stack.TryPop(out poppedValue))
                        {
                        global::Tsonic.JSRuntime.console.log("Stack pop:", poppedValue);
                        }
                    var queue = new Queue<string>();
                    queue.Enqueue("first");
                    queue.Enqueue("second");
                    queue.Enqueue("third");
                    global::Tsonic.JSRuntime.console.log("Queue size:", queue.size());
                    string dequeuedValue = "";
                    if (queue.TryDequeue(out dequeuedValue))
                        {
                        global::Tsonic.JSRuntime.console.log("Queue dequeue:", dequeuedValue);
                        }
                    var list = new LinkedList<double>();
                    list.append(10);
                    list.append(20);
                    list.prepend(5);
                    global::Tsonic.JSRuntime.console.log("List as array:", list.toArray());
                    }
            }
}